### PR TITLE
Use combined W matrix for primary and secondary data in IQN

### PR DESCRIPTION
**Pull request by Fujikawas**
## Main changes of this PR
The secondary data is handled together with primary data in the `W` matrix both in IQN-ILS and IQN-IMVJ method.

## Motivation and additional information
Relating to #1996 

## To do

* [x] Check if the `secondaryDataIDs` and `_values` with only primary data are still needed; -> Yes, for computing `V`, warn zero update of `V`;
* [x] The preconditioning of `Wtil` should still be done, currently commented out;
* [x] With the tests with secondary data in #2006, this method could work. Still the QNTest9EmptyPartition which is to test SVD-restart failed. Other tests with SVD could work.

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [x] I added a test to cover the proposed changes in our test suite.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [x] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?

### CODE REVIEW

File: src/acceleration/BaseQNAcceleration.hpp [hunk 1]
```diff
@@ -161,6 +161,9 @@ class BaseQNAcceleration : public Acceleration {
   /// Data IDs of data not involved in IQN coefficient computation.
   std::vector<int> _secondaryDataIDs;
 
+  /// Data IDs of all coupling data.
+  std::vector<int> _cplDataIDs;
+
```
Comment (uekerman) [inline src/acceleration/BaseQNAcceleration.hpp:166]: Is `_dataIDs` then only primary data?
If yes, I would suggest to rename: `_dataIDs` -> `_primaryDataIDs` and `cplDataIDs` -> `_dataIDs`.

File: src/acceleration/BaseQNAcceleration.hpp [hunk 2]
```diff
@@ -274,6 +277,12 @@ class BaseQNAcceleration : public Acceleration {
   /// @brief Concatenation of all (old) coupling data involved in the QN system.
   Eigen::VectorXd _oldValues;
 
+  /// @brief Concatenation of all primary and secondary data involved in the QN system.
+  Eigen::VectorXd _cplValues;
+
+  /// @brief Concatenation of all old primary and secondary data involved in the QN system.
+  Eigen::VectorXd _oldCplValues;
+
```
Comment (uekerman) [inline src/acceleration/BaseQNAcceleration.hpp:285]: We could also rename here to make things more explicit, see above.

**Review by uekerman — COMMENTED** [review]
Looks promising!
I guess the SVD changes could go to a separate refactoring PR.

File: src/acceleration/BaseQNAcceleration.cpp [hunk 3]
```diff
@@ -95,25 +95,41 @@ void BaseQNAcceleration::initialize(
                   "Quasi-Newton acceleration does not yet support using data from all substeps. Please set substeps=\"false\" in the exchange tag of data \"{}\".", data->getDataName());
   }
 
-  size_t              entries = 0;
+  size_t              entries     = 0;
+  size_t              cplDataSize = 0;
```
Comment (uekerman) [inline src/acceleration/BaseQNAcceleration.cpp:99]:
```suggestion
  size_t              primaryDataSize = 0;
  size_t              dataSize = 0;
```

File: src/acceleration/BaseQNAcceleration.cpp [hunk 4]
```diff
@@ -149,24 +166,15 @@ void BaseQNAcceleration::initialize(
     }
 
     // test that the computed number of unknown per proc equals the number of entries actually present on that proc
-    size_t unknowns = _dimOffsets[utils::IntraComm::getRank() + 1] - _dimOffsets[utils::IntraComm::getRank()];
-    PRECICE_ASSERT(entries == unknowns, entries, unknowns);
+    // size_t unknowns = _dimOffsets[utils::IntraComm::getRank() + 1] - _dimOffsets[utils::IntraComm::getRank()];
+    // PRECICE_ASSERT(entries == unknowns, entries, unknowns); //?
```
Comment (uekerman) [inline src/acceleration/BaseQNAcceleration.cpp:170]: This is a tricky point. Should `_dimOffsets` include secondary data or not? Probably makes a difference for IMVJ. Let's try to also add a test with secondary data that runs in parallel.

File: src/acceleration/BaseQNAcceleration.cpp [hunk 5]
```diff
@@ -275,18 +286,19 @@ void BaseQNAcceleration::updateDifferenceMatrices(
...
-  PRECICE_ASSERT(_oldValues.size() == _oldXTilde.size(), _oldValues.size(), _oldXTilde.size());
-  PRECICE_ASSERT(_residuals.size() == _oldXTilde.size(), _residuals.size(), _oldXTilde.size());
+  PRECICE_ASSERT(_oldPrimaryResiduals.size() == _oldPrimaryXTilde.size(), _oldPrimaryResiduals.size(), _oldPrimaryXTilde.size());
+  PRECICE_ASSERT(_primaryValues.size() == _oldPrimaryXTilde.size(), _primaryValues.size(), _oldPrimaryXTilde.size());
+  PRECICE_ASSERT(_oldPrimaryValues.size() == _oldPrimaryXTilde.size(), _oldPrimaryValues.size(), _oldPrimaryXTilde.size());
+  PRECICE_ASSERT(_primaryResiduals.size() == _oldPrimaryXTilde.size(), _primaryResiduals.size(), _oldPrimaryXTilde.size());
 
   // assume data structures associated with the LS system can be updated easily.
 
   // scale data values (and secondary data values)
```
Comment (uekerman) [inline src/acceleration/BaseQNAcceleration.cpp:300]: Is this comment correct? I don't think that there is any scaling in there.

File: src/acceleration/BaseQNAcceleration.cpp [hunk 6]
```diff
@@ -298,16 +310,15 @@ void BaseQNAcceleration::performAcceleration(
...
+    _oldPrimaryXTilde    = _primaryValues;    // Store x tilde of primary data
+    _oldXTilde           = _values;           // Store x tilde of primary and secondary data
+    _oldPrimaryResiduals = _primaryResiduals; // Store current residual of primary data
 
     // Perform constant relaxation
     // with residual: x_new = x_old + omega * res
     _residuals *= _initialRelaxation;
-    _residuals += _oldValues;
-    _values = _residuals;
+    _values = _oldValues + _residuals;
```
Comment (uekerman) [inline src/acceleration/BaseQNAcceleration.cpp:320]: Is this correct? Now `_residuals` has a different value.

File: src/acceleration/impl/ParallelMatrixOperations.hpp [hunk 7]
```diff
@@ -136,7 +136,6 @@ class ParallelMatrixOperations {
 
     PRECICE_ASSERT(_needCyclicComm);
     PRECICE_ASSERT(leftMatrix.cols() == q, leftMatrix.cols(), q);
-    PRECICE_ASSERT(leftMatrix.rows() == rightMatrix.cols(), leftMatrix.rows(), rightMatrix.cols());
```
Comment (uekerman) [inline src/acceleration/impl/ParallelMatrixOperations.hpp:139]: Removing this doesn't feel good.

File: src/acceleration/impl/Preconditioner.hpp [hunk 8]
```diff
@@ -58,16 +58,22 @@ class Preconditioner {
   {
     PRECICE_TRACE();
     if (transpose) {
-      PRECICE_ASSERT(M.cols() == (int) _weights.size(), M.cols(), _weights.size());
-      for (int i = 0; i < M.cols(); i++) {
+      PRECICE_WARN_IF(M.cols() < (int) _weights.size(), "Matrix and weights size mismatch, only part of the preconditioning weights will be used.");
+      PRECICE_WARN_IF(M.cols() > (int) _weights.size(), "Matrix and weights size mismatch, only part of the matrix will be preconditioned.");
```
Comment (uekerman) [inline src/acceleration/impl/Preconditioner.hpp:62]: Let's find a better solution here. Such warnings will confuse 99% of users.

File: src/acceleration/impl/Preconditioner.hpp
Comment (uekerman) [inline src/acceleration/impl/Preconditioner.hpp:1]: Could we add unit tests to cover these changes?

File: src/acceleration/impl/SVDFactorization.hpp
Comment (uekerman) [inline src/acceleration/impl/SVDFactorization.hpp:1]: I still think it would be better to do these changes in a separate PR before this one. And add unit tests for the new functionality.

**Review by uekerman — CHANGES_REQUESTED** [review]
Nice to see how these changes reduce code duplication :+1:

Still a bit of work to do, especially for parallel runs I guess, see below.

Requires a changelog as we now have support of secondary data for IMVJ

File: src/acceleration/BaseQNAcceleration.cpp (hunk 6 above)
Comment (Fujikawas) [reply src/acceleration/BaseQNAcceleration.cpp:320]: From the test results I didn't see difference. If you mean that the `_residuals` is not updated here, I see that the `_residuals` would be calculated from `_values` and `_oldValues` by updating differences and is not stored like `_values`, so the step here still works.

File: src/acceleration/BaseQNAcceleration.cpp (hunk 5 above)
Comment (Fujikawas) [reply src/acceleration/BaseQNAcceleration.cpp:300]: Only concatenate happens until the next comment

File: src/acceleration/BaseQNAcceleration.hpp [hunk 9]
```diff
@@ -90,7 +90,7 @@ class BaseQNAcceleration : public Acceleration {
    */
   virtual std::vector<int> getDataIDs() const
   {
-    return _dataIDs;
+    return _primaryDataIDs;
   }
```
[the remark this answers is not in the document]
Comment (Fujikawas) [reply src/acceleration/BaseQNAcceleration.hpp:94]: For the whole acceleration method? Other under-relaxation methods also use this function

File: src/acceleration/impl/ParallelMatrixOperations.hpp (hunk 7 above)
Comment (Fujikawas) [reply src/acceleration/impl/ParallelMatrixOperations.hpp:139]: In the comment it says that ` For multiplication W_til * Z = J, p = r = n_global, q = m`, which is not the case now. I'm not so sure why this assertion is necessary for the operation unless these two variables are mixed used. Need to check.

File: src/acceleration/BaseQNAcceleration.cpp (hunk 4 above)
Comment (Fujikawas) [reply src/acceleration/BaseQNAcceleration.cpp:170]: Since it's used to initialize the size of `Wtil`, I would say that the `_dimOffsets` should include the secondary data. But I'm a little confused with this function now: https://github.com/precice/precice/blob/8458db5189c7f4c9c41551cac55eb957ce46af80/src/acceleration/BaseQNAcceleration.cpp#L626-L632
In the extended tests, the parallel tests with IMVJ include also secondary data. They could work with this implementation here(secondary data included in the `_dimOffsets`). This assert should be modified a bit then.

File: src/acceleration/BaseQNAcceleration.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (uekerman) [reply src/acceleration/BaseQNAcceleration.hpp:94]: For Aitken, we in principle also have the primary vs secondary story.

[the remark this answers is not in the document]
Comment (uekerman) [reply src/acceleration/BaseQNAcceleration.hpp:94]: Why not `getPrimaryDataIDs()`? Otherwise, this might be confusing. And makes sense for all acceleration methods, no?

File: src/acceleration/BaseQNAcceleration.cpp (hunk 5 above)
Comment (uekerman) [reply src/acceleration/BaseQNAcceleration.cpp:309]: But then the comment is misleading, no?

File: src/acceleration/test/ParallelMatrixOperationsTest.cpp
Comment (uekerman) [inline src/acceleration/test/ParallelMatrixOperationsTest.cpp:1]: Could it make sense to add more tests here?

**Review by uekerman — COMMENTED** [review]
I did not check all implementation details, but I have a good feeling in general.
Would it make sense to also add unit tests?

File: src/acceleration/BaseQNAcceleration.cpp (hunk 5 above)
Comment (uekerman) [reply src/acceleration/BaseQNAcceleration.cpp]: now, it says only `scale data values`, but there is still no scaling happening, no? :)

File: src/acceleration/test/ParallelMatrixOperationsTest.cpp
Comment (uekerman) [reply src/acceleration/test/ParallelMatrixOperationsTest.cpp:1]: looks good! please add a quick comment how you created the test

**Review by uekerman — APPROVED** [review]
